optuna/samplers/_nsgaiii.py

```diff
@@ -0,0 +1,637 @@
+    # not seem necessary to calculate distance from all reference points.
+
+    # TODO(Shinichi) Normalize reference_points in constructor to remove reference_point_norms.
+    # In addition, the minimum distance from each reference point can be replaced with maximum
+    # inner product between the given individual and each normalized reference points.
+
+    # distance_from_reference_lines is a ndarray of shape (n, p), where n is the size of the
+    # population and p is the number of reference points. Its (i,j) entry keeps distance between
+    # the i-th individual values and the j-th reference line.
+    reference_point_norm_squared = np.linalg.norm(reference_points, axis=1) ** 2
+    perpendicular_vectors_to_reference_lines = np.einsum(
+        "ni,pi,p,pm->npm",
+        objective_matrix,
+        reference_points,
+        1 / reference_point_norm_squared,
+        reference_points,
+    )
+    distance_from_reference_lines = np.linalg.norm(
+        objective_matrix[:, np.newaxis, :] - perpendicular_vectors_to_reference_lines,
+        axis=2,
+    )
+    closest_reference_points: np.ndarray = np.argmin(distance_from_reference_lines, axis=1)
+    distance_reference_points: np.ndarray = np.min(distance_from_reference_lines, axis=1)
+
+    return closest_reference_points, distance_reference_points
+
+
+def _preserve_niche_individuals(
+    target_population_size: int,
+    elite_population_num: int,
+    population: list[FrozenTrial],
+    closest_reference_points: np.ndarray,
+    distance_reference_points: np.ndarray,
+    rng: np.random.RandomState,
+) -> list[FrozenTrial]:
+    """Determine who survives form the borderline front
+
+    Who survive form the borderline front is determined according to the sparsity of each closest
+    reference point. The algorithm picks a reference point from those who have the least neighbors
+    in elite population and adds one of borderline front member who has the same closest reference
+    point.
+
+    Args:
+        target_population_size:
+            The number of individuals to select.
+        elite_population_num:
+            The number of individuals which are already selected as the elite population.
+        population:
+            List of all the trials in the current surviving generation.
+        distance_reference_points:
+            A ``numpy.ndarray`` with rows of generation size. Each row is the distance from the
+            corresponding individual to the closest reference point.
+        closest_reference_points:
+            A ``numpy.ndarray`` with rows of generation size. Each row is the index of the closest
+            reference point to the corresponding individual.
+        rng:
+            Random number generator.
+
+    Returns:
+        A list of trials which are selected as the next generation.
+    """
+    if len(population) < target_population_size:
+        raise IndexError(
```

Comment:
Could you explain why you choose `IndexError`? `ValueError` may be suitable since input values are invalid.

Reply:
I fixed it. Thank you for the indication.